File: cpp/leetcode/076.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int m = s.size(), n = t.size(), miss = t.size();
        vector<int> need(128, 0);
        for (int i = 0; i < n; i++) need[t[i]]++;
        int width = INT_MAX, start = 0, left = 0;
        for(int i = 0; i <= s.size(); ) {
            if(miss) {
                if (i == s.size()) break;
                if (--need[s[i++]] >= 0) miss--;
            } else {
                if (i - start < width) {
                    width = i - start;
                    left = start;
                }
                if (++need[s[start++]] > 0) miss ++;
            }
        }
        return width == INT_MAX? "" : s.substr(left, width);
    }
};

File: python/leetcode/076.py
import collections
class Solution:
    def minWindow(self, s, t):
        need, missing = collections.Counter(t), len(t)
        i, l, r = 0, 0, 0
        for idx, val in enumerate(s, 1):
            missing -= need[val] > 0  # need[val] <= 0, will get false == 0
            need[val] -= 1
            if not missing:
                while i < idx and need[s[i]] < 0:
                    need[s[i]] += 1
                    i += 1
                if not r or idx - i <= r - l:
                    l, r = i, idx
            print need, missing, l, r
        return s[l:r]


if __name__ == "__main__":
    test = Solution()
    print test.minWindow("ADOBECODEBANC", "ABC")